vm/systemSmartContracts/governance.go

```diff
@@ -288,521 +499,599 @@ func (g *governanceContract) whiteListProposal(args *vmcommon.ContractCallInput)
-	switch voteValue {
-	case "yes":
-		generalProposal.Yes += numVotes
-	case "no":
-		generalProposal.No += numVotes
-	case "veto":
-		generalProposal.Veto += numVotes
-	case "dontCare":
-		generalProposal.DontCare += numVotes
+// getLock returns the lock nonce for a voter
+func (g *governanceContract) getLock(voter []byte, voteType VoteType, proposalReferance []byte) *big.Int {
+	prefix := []byte(validatorLockPrefix)
+	if voteType == Account {
+		prefix = append([]byte(accountLockPrefix), proposalReferance...)
 	}
+	lockKey := append(prefix, voter...)
+
+	lock := g.eei.GetStorage(lockKey)
+
+	return big.NewInt(0).SetBytes(lock)
 }
 
-func (g *governanceContract) saveVoteValue(proposal []byte, voter []byte, voteData *VoteData) error {
-	key := append(proposal, voter...)
-	marshaledData, err := g.marshalizer.Marshal(voteData)
-	if err != nil {
-		return err
+// proposalContainsVoter iterates through all the votes on a proposal and returns if it already contains a
+//  vote from a certain address
+func (g *governanceContract) proposalContainsVoter(proposal *GeneralProposal, voteKey []byte) bool {
+	for _, vote := range proposal.Votes {
+		if bytes.Equal(vote, voteKey) {
+			return true
+		}
 	}
 
-	g.eei.SetStorage(key, marshaledData)
-	return nil
+	return false
 }
 
-func (g *governanceContract) getOrCreateVoteData(proposal []byte, voter []byte) (*VoteData, error) {
-	voteData := &VoteData{}
-	key := append(proposal, voter...)
-	marshaledData := g.eei.GetStorage(key)
-	if len(marshaledData) == 0 {
-		return voteData, nil
+// computeVotingPower returns the voting power for a value. The value can be either a balance or
+//  the staked value for a validator
+func (g *governanceContract) computeVotingPower(value *big.Int) (*big.Int, error) {
+	if value.Cmp(big.NewInt(0)) == -1 {
+		return nil, fmt.Errorf("cannot compute voting power on a negative value")
 	}
 
-	err := g.marshalizer.Unmarshal(voteData, marshaledData)
-	if err != nil {
-		return nil, err
+	return big.NewInt(0).Sqrt(value), nil
+}
+
+// getDelegatedToAddress looks into the arguments passed and returns the optional delegatedTo address
+func (g *governanceContract) getDelegatedToAddress(args *vmcommon.ContractCallInput) ([]byte, error) {
+	if len(args.Arguments) < 4 {
+		return make([]byte, 0), nil
+	}
+	if len(args.Arguments[3]) != len(args.CallerAddr) {
+		return nil, fmt.Errorf("%s: %s", vm.ErrInvalidArgument, "invalid delegator address length")
 	}
 
-	return voteData, nil
+	return args.Arguments[2], nil
 }
 
-func (g *governanceContract) getOrCreateValidatorData(address []byte, numNodes int32) (*ValidatorData, error) {
-	validatorData := &ValidatorData{
-		Delegators: make([]*VoterData, 1),
-		NumNodes:   numNodes,
+// isValidVoteString checks if a certain string represents a valid vote string
+func (g *governanceContract) isValidVoteString(vote string) bool {
+	switch vote {
+	case "yes":
```

Comment:
Good catch, thanks. I've added return instead of `fallthrough` since it's also one liner and seems more explicit. Also added constants instead of magic strings

Reply:
this can be better simplified and look better:
```GO
func (g *governanceContract) isValidVoteString(vote string) bool {
    switch vote {
    case yesString, noString, vetoString:
	    return true
    default:
            return false
    }
}
```

Reply:
Better, changed it 👍